phase1/gce/configure-vm-kubeadm.sh

```diff
@@ -3,30 +3,52 @@
 TOKEN=$(get_metadata "k8s-kubeadm-token")
 KUBEADM_VERSION=$(get_metadata "k8s-kubeadm-version")
 KUBERNETES_VERSION=$(get_metadata "k8s-kubernetes-version")
-
+KUBELET_VERSION=$(get_metadata "k8s-kubelet-version")
 curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
 
-if [[ "${KUBEADM_VERSION}" == stable ]]; then
+if [[ "${KUBELET_VERSION}" == stable ]]; then
   cat <<EOF > /etc/apt/sources.list.d/kubernetes.list
   deb http://apt.kubernetes.io/ kubernetes-xenial main
 EOF
-
   apt-get update
+  # kubeadm is installed with the kubelet so that the 
+  # kubelet has the configuration at a matching version
   apt-get install -y kubelet kubeadm kubectl kubernetes-cni
-elif [[ "${KUBEADM_VERSION}" == "gs://"* ]]; then
+elif [[ "${KUBELET_VERSION}" == "gs://"* ]]; then
   TMPDIR=/tmp/k8s-debs
   mkdir $TMPDIR
-  gsutil rsync "${KUBEADM_VERSION}" $TMPDIR
+  gsutil rsync "${KUBELET_VERSION}" $TMPDIR
+  # kubeadm is installed with the kubelet so that the 
+  # kubelet has the configuration at a matching version
   dpkg -i $TMPDIR/{kubelet,kubeadm,kubectl,kubernetes-cni}.deb || echo Ignoring expected dpkg failure
   apt-get install -f -y
   systemctl enable kubelet
   systemctl start kubelet
   rm -rf $TMPDIR
 else
-  echo "Don't know how to handle version: $KUBEADM_VERSION"
+  echo "Don't know how to handle kubelet version: $KUBELET_VERSION"
   exit 1
 fi
 
+if [[ "${KUBEADM_VERSION}" != "${KUBELET_VERSION}" ]]; then
+  if [[ "${KUBEADM_VERSION}" == stable ]]; then
+    # Cannot install packages as they will update the 
+    # kubelet configuration to a version that does not match
+    # the installed kubelet
+    echo "Kubeadm version of 'stable' is not supported with kubelet version that is not also 'stable'."
+    exit 1
+  elif [[ "${KUBEADM_VERSION}" == "gs://"* ]]; then
```

Comment:
Similarly, I would expect there to be separate checks for either KUBEADM_VERSION or KUBELET_VERSION to match the gs:// scheme, which would then trigger the `gcloud rsync` to pull the artifacts and `dpkg -i kubelet.deb` or `dpkg -i kubeadm.deb`.

The way this is written, if kubeadm is a gs:// link and kubelet is "stable," then both kubeadm and kubelet will be installed from the stable repository, and only the kubeadm binary will be replaced from the correct location. The kubeadm .deb package contains more than just this binary, so this wouldn't do the correct thing if, say, the `10-kubeadm.conf` configuration file were different between the two sources.

Reply:
The avoidance of using packages for updating kubeadm to another version is to avoid updating the kubelet config (eg. the stuff in 10-kubeadm.conf) to an incompatible future version. There is a bit of catch-22 involved in this.

If you know of a better way to update kubeadm without changing the kubelet config, please do let me know.